src/Init/Data/Iterators/Combinators/Monadic/Attach.lean

```diff
@@ -0,0 +1,136 @@
+    obtain ⟨_, hs, step, h', rfl⟩ := h
+    cases step using PlausibleIterStep.casesOn
+    · simp only [IterStep.successor, modifyStep, Option.some.injEq] at hs
+      simp only [← hs]
+      exact ⟨_, rfl, ‹_›⟩
+    · simp only [IterStep.successor, modifyStep, Option.some.injEq] at hs
+      simp only [← hs]
+      exact ⟨_, rfl, ‹_›⟩
+    · simp [IterStep.successor, modifyStep, reduceCtorEq] at hs
+
+instance Attach.instFinite {α β : Type w} {m : Type w → Type w'} [Monad m]
+    [Iterator α m β] [Finite α m] {P : β → Prop} : Finite (Attach α m P) m :=
+  .of_finitenessRelation instFinitenessRelation
+
+def Attach.instProductivenessRelation {α β : Type w} {m : Type w → Type w'} [Monad m]
+    [Iterator α m β] [Productive α m] {P : β → Prop} :
+    ProductivenessRelation (Attach α m P) m where
+  rel := InvImage WellFoundedRelation.rel fun it => it.internalState.inner.finitelyManySkips
+  wf := InvImage.wf _ WellFoundedRelation.wf
+  subrelation {it it'} h := by
+    apply Relation.TransGen.single
+    simp_wf
+    obtain ⟨step, hs⟩ := h
+    cases step using PlausibleIterStep.casesOn
+    · simp [modifyStep] at hs
+    · simp only [modifyStep, IterStep.skip.injEq] at hs
+      simp only [← hs]
+      assumption
+    · simp [modifyStep] at hs
+
+instance Attach.instProductive {α β : Type w} {m : Type w → Type w'} [Monad m]
+    [Iterator α m β] [Productive α m] {P : β → Prop} :
+    Productive (Attach α m P) m :=
+  .of_productivenessRelation instProductivenessRelation
+
+instance Attach.instIteratorCollect {α β : Type w} {m : Type w → Type w'} [Monad m] [Monad n]
+    {P : β → Prop} [Iterator α m β] :
+    IteratorCollect (Attach α m P) m n :=
+  .defaultImplementation
+
+instance Attach.instIteratorCollectPartial {α β : Type w} {m : Type w → Type w'} [Monad m]
+    [Monad n] {P : β → Prop} [Iterator α m β] :
+    IteratorCollectPartial (Attach α m P) m n :=
+  .defaultImplementation
+
+instance Attach.instIteratorLoop {α β : Type w} {m : Type w → Type w'} [Monad m]
+    [Monad n] {P : β → Prop} [Iterator α m β] [MonadLiftT m n] :
+    IteratorLoop (Attach α m P) m n :=
+  .defaultImplementation
+
+instance Attach.instIteratorLoopPartial {α β : Type w} {m : Type w → Type w'} [Monad m]
+    [Monad n] {P : β → Prop} [Iterator α m β] [MonadLiftT m n] :
+    IteratorLoopPartial (Attach α m P) m n :=
+  .defaultImplementation
+
+instance {α β : Type w} {m : Type w → Type w'} [Monad m]
+    {P : β → Prop} [Iterator α m β] [IteratorSize α m] :
+    IteratorSize (Attach α m P) m where
+  size it := IteratorSize.size it.internalState.inner
+
+instance {α β : Type w} {m : Type w → Type w'} [Monad m]
+    {P : β → Prop} [Iterator α m β] [IteratorSizePartial α m] :
+    IteratorSizePartial (Attach α m P) m where
+  size it := IteratorSizePartial.size it.internalState.inner
+
+end Types
+
+/--
+“Attaches” individual proofs to an iterator of values that satisfy a predicate `P`, returning a list
+of iterators in the corresponding subtype `{ x // P x }`.
```

Comment:
I'm confused by this docstring. Where exactly does the list come in?

Reply:
I mistyped, so your confusion is warranted!